### DISCUSSION TITLE
Recommendations on when to flush?

### DISCUSSION THREAD

**Discussion by johanneslumpe** [mikro-orm/mikro-orm#3838] [Q&A] [answered]
I'm looking into mikro-orm for a project and I'm curious when the best time for flushing would be? I read through https://github.com/mikro-orm/mikro-orm/discussions/2416 which provided some ideas, but I'm still not fully certain. I could flush at the end of a controller/use case. In GraphQL that would be at the end of a mutation, before data gets loaded again. Does this make sense? I expect that data must be flushed before any data can be returned to the client to ensure that there are no conflicts, e.g. when flushing versioned entities. Am I correct to assume that a centralized place in code to flush doesn't really make sense and that you do want to flush for each use case?

**Comment by B4nan** [comment, marked as answer]
As with everything, it depends :] Flushing will have a small performance overhead as it always checks all entities in context for changes. If you have an endpoint that works with thousands of loaded entities, it's not a good idea to just blindly flush, as it will hurt the performance. 

Personally, I'd flush only where I mutate the state, where I expect some queries from the flush operation. But if you know you won't have a huge amount of loaded entities in one context, flushing in some middleware might work just fine. I don't think many people do it, at least I don't see any issues in that direction.

**Reply by johanneslumpe** [reply]
I see! For a service pattern you’d want to flush outside of the services though I’d imagine if you want an all or nothing write within a transaction, correct? And a follow up: you must flush _somewhere_ to persist changes in a context, correct? I want to make sure I’m not misunderstanding how flushing works.

**Reply by B4nan** [reply]
Yes, but flush can also happen automatically https://mikro-orm.io/docs/unit-of-work#flush-modes

**Reply by johanneslumpe** [reply]
Great thanks! I’ll read up more on it and also run some tests to ensure I’m fully understanding. Appreciate the help!
